website/public_adventures.py

```diff
@@ -31,67 +31,168 @@ def __init__(self, db: Database, achievements: Achievements):
-            # Perhaps we could make this name#creator!
-            if included.get(adventure["name"]):
-                continue
-            public_profile = self.db.get_public_profile_settings(adventure.get('creator'))
-            included[adventure["name"]] = True
-
-            content = safe_format(adventure.get('formatted_content', adventure['content']),
-                                  **hedy_content.KEYWORDS.get(g.keyword_lang))
-            current_adventure = {
-                "id": adventure.get("id"),
-                "name": adventure.get("name"),
-                "short_name": adventure.get("name"),
-                "author": adventure.get("author", adventure["creator"]),
-                "creator": adventure.get("creator"),
-                "creator_public_profile": public_profile,
-                "date": utils.localized_date_format(adventure.get("date")),
-                "level": adventure.get("level"),
-                "levels": adventure.get("levels"),
-                "language": adv_lang,
-                "cloned_times": adventure.get("cloned_times"),
-                "tags": adv_tags,
-                "text": content,
-                "is_teacher_adventure": True,
-            }
-
-            # save adventures for later usage.
-            for _level in adventure.get("levels", [adventure.get("level")]):
-                _level = int(_level)
-                if self.adventures.get(_level):
-                    self.adventures[_level].append(current_adventure)
-                else:
-                    self.adventures[_level] = [current_adventure]
-
-            available_levels = adventure["levels"] if adventure.get("levels") else [adventure["level"]]
-            self.customizations["available_levels"].update([int(adv_level) for adv_level in available_levels])
-
     @route("/", methods=["GET"])
     @route("/filter", methods=["POST"])
     @requires_teacher
-    def filtering(self, user, index_page=False):
+    def filtering(self, user):
+        level = request.args["level"] if request.args.get("level") else "1"
+        language = request.args.get("lang", g.lang)
+        tag = request.args.get("tag", "")
+        search = request.form.get("search", request.args.get("search", ""))
+        print('\n\n\n', level, language, tag, search, '\n\n\n')
+
+        available_languages = set()
+        available_levels = set()
+        available_tags = set()
+
+        # Get all possible filters
+        lang_level_filters = self.db.get_public_adventures_filters({"field": "lang#level", })
+        for record in lang_level_filters:
+            lang, _level = record.get("value", "#").split("#")
+            available_levels.update([_level])
+            if level == _level:
+                available_languages.update([lang])
+        customizations = {"available_levels": available_levels}
+
+        tag_filters = self.db.get_public_adventures_filters({"field": "tag"})
```

Comment:
You could have gotten all of these in one go using a table scan. Saving roundtrips is a good idea.